### DISCUSSION TITLE
set-cookie outside of loader

### DISCUSSION THREAD

**Discussion by dusty** [remix-run/remix#2166] [Q&A] [answered]
Looking through the FAQ (https://remix.run/docs/en/v1/pages/faq), I found this snippet for creating a function that will protect routes.

Now, say I added a refreshToken type of action to that function, that grabs the current token/refreshToken that is stored in the cookie.  When I call that from the loader, is there any way to set-cookie without having to do it manually on every loader?

The main reason I would like this, is because I don't know if I'm going to get a new token in the requireUserSession method, so I would have to set-cookie every time on every loader in the app.  

Is it doable?

```
export async function requireUserSession(request) {
  // get the session
  const cookie = request.headers.get("cookie");
  const session = await getSession(cookie);

  const token = session.get('token')
  const refreshToken = session.get('refreshToken')
  if (isValid(token)) return session

  // if we don't have a refresh, bail
  if (!isValid(refreshToken)) throw redirect('/login')

  // lets get a new one
  try {
    const res = await fetch('/my/awesome/api', {headers: {'authorization': `Bearer ${refreshToken}`})
    session.set('token', res.token)
    session.set('refreshToken', res.refreshToken)
    await storage.commitSession(session)

    // can I magically handle this here so I don't have to do it in the loader?
    // request.response.headers.set('set-cookie', ...)

    return session
  } catch (err) {
    throw redirect('/login')
  }
}
```

```
export const loader: LoaderFunction = ({request}) => {
  const session = requireUserSession(request)
  const otherThing = fetchMyApi('/some/route', session.get('token'))
  const thatThing = fetchMyApi('/some/other', session.get('token'))

  // I'd prefer to just return { otherThing, thatThing }
  return json({otherThing, thatThing}, { headers: { 'set-cookie': session.toString()}))
}

export default function Thing() {
  const {otherThing, thatThing} = useLoaderData()
}
```

**Comment by kiliman** [comment, marked as answer]
Hmm.. there's no "easy" way to do this, as you're still dealing with JavaScript. Somehow you need to "magically" send arguments from one function `requireUserSession` to another `json` with no manual code.

Here are a couple of possible "solutions":

1) In `requireUserSession`, if you have to get a new token, instead of returning directly, just `throw redirect(request.url, { headers: { 'set-cookie': ... })`. This will send the new cookie and the browser will refetch the loader at which point you now have the correct token.

2) In `requireUserSession`, instead of simply returning the session object, you can return a new `json` function that already has the headers.

BTW: This is *browser* code, so may not work 100%, but you get the gist.
```ts
import { json as remix_json } from 'remix'

export async function requireUserSession(request: Request) {
  // get the session
  const cookie = request.headers.get('cookie')
  const session = await getSession(cookie)

  const headers = new Headers()
  // get new token and store in our headers
  headers.append('set-cookie', 'token=12345')

  function json(data: any, init?: number | ResponseInit) {
    if (typeof init === 'number') {
      init = { status: init, headers }
    } else if (init) {
      // merge headers with user provided headers (via closure)
      if (init.headers) {
        const entries =
          init.headers instanceof Headers
            ? Array.from(init.headers.entries())
            : Object.entries(init.headers)

        entries.forEach(([key, value]) => headers.append(key, value))
      }
      init.headers = headers // replace with our headers
    }
    return remix_json(data, init)
  }

  return { session, json }
}
```
```ts
// route.tsx
export const loader: LoaderFunction = ({request}) => {
  const { session, json } = requireUserSession(request)
  const otherThing = fetchMyApi('/some/route', session.get('token'))
  const thatThing = fetchMyApi('/some/other', session.get('token'))

  // you can now return your data, and optionally set headers, etc.
  return json({otherThing, thatThing})
}

export default function Thing() {
  const {otherThing, thatThing} = useLoaderData()
}
```

**Reply by dusty** [reply]
I really like both of those answers.

The first one is super interesting, sending a redirect to the same url in order to set the new headers, which then causes the loader to reload.  

I ended up going down a path of something similar to your second answer, which made it little easier to reason about when I was returning the loader.  Instead of returning a json function, I was just returning the generated headers `{ headers: {.....`

But, I like the idea of returning a function because its pretty specific to to what the responsibility of this requireUserSession method.

Thanks for the insights!

```
const { stuff, headers } = await requireUserSession(request)
return json({thing}, headers)
```
